WordPressKit/HTTPClient.swift

```diff
@@ -9,37 +9,92 @@ struct HTTPAPIResponse<Body> {
+    /// parameter, which must satisify following requirements:
+    /// - `totalUnitCount` must not be zero.
+    /// - `completedUnitCount` must be zero.
+    /// - It's used exclusivity for tracking the HTTP request overal progress: No children in its progress tree.
+    /// - `cancellationHandler` must be nil. You can call `fulfillingProgress.cancel()` to cancel the ongoing HTTP request.
+    ///
+    ///  Upon completion, the HTTP request's progress fulfills the `fulfillingProgress`.
+    ///
+    /// - Parameters:
+    ///   - builder: A `HTTPRequestBuilder` instance that represents an HTTP request to be sent.
+    ///   - acceptableStatusCodes: HTTP status code ranges that are considered a successful response. Responses with
+    ///         a status code outside of these ranges are returned as a `WordPressAPIResult.unacceptableStatusCode` instance.
+    ///   - parentProgress: A `Progress` instance that will be used as the parent progress of the HTTP request's overall
+    ///         progress. See the function documentation regarding requirements on this argument.
+    ///   - errorType: The concret endpoint error type.
     func perform<E: LocalizedError>(
         request builder: HTTPRequestBuilder,
         acceptableStatusCodes: [ClosedRange<Int>] = [200...299],
+        fulfillingProgress parentProgress: Progress? = nil,
         errorType: E.Type = E.self
     ) async -> WordPressAPIResult<HTTPAPIResponse<Data>, E> {
+        if let parentProgress {
+            assert(parentProgress.completedUnitCount == 0 && parentProgress.totalUnitCount > 0, "Invalid parent progress")
+            assert(parentProgress.cancellationHandler == nil, "The progress instance's cancellationHandler property must be nil")
+        }
+
         guard let request = try? builder.build() else {
             return .failure(.requestEncodingFailure)
         }
 
-        let result: (Data, URLResponse)
-        do {
-            result = try await data(for: request)
-        } catch {
-            if let urlError = error as? URLError {
-                return .failure(.connection(urlError))
-            } else {
-                return .failure(.unknown(underlyingError: error))
+        return await withCheckedContinuation { continuation in
+            let task = dataTask(with: request) { data, response, error in
+                let result: WordPressAPIResult<HTTPAPIResponse<Data>, E> = Self.parseResponse(
+                    data: data,
+                    response: response,
+                    error: error,
+                    acceptableStatusCodes: acceptableStatusCodes
+                )
+
+                continuation.resume(returning: result)
             }
-        }
+            task.resume()
 
-        let (body, response) = result
+            if let parentProgress, parentProgress.totalUnitCount > parentProgress.completedUnitCount {
+                let pending = parentProgress.totalUnitCount - parentProgress.completedUnitCount
```

Comment:
Do we need this math given the assert above for `completedUnitCount == 0`?

Reply:
No, I don't think it's needed. But, in my head, the code expresses the idea more explicitly: letting the `task.progress` fill `parentProgress` without overflowing.

Reply:
Fair 👍